std/stdio.d

```diff
@@ -2603,6 +2606,213 @@ See $(LREF byChunk) for an example.
         return LockingTextWriter(this);
     }
 
+    // An output range which optionally locks the file and puts it into
+    // binary mode (similar to rawWrite). Because it needs to restore
+    // the file mode on destruction, it is RefCounted on Windows.
+    struct BinaryWriterImpl(bool locking)
+    {
+    private:
+        FILE* fps;
+        string name;
+
+        version(Windows)
+        {
+            int fd, oldMode;
+            version(DIGITAL_MARS_STDIO)
+                ubyte oldInfo;
+        }
+
+    package:
+        this(ref File f)
+        {
+            import std.exception : enforce;
+
+            enforce(f._p && f._p.handle);
+            name = f._name;
+            fps = f._p.handle;
+            static if (locking)
+                FLOCK(fps);
+
+            version(Windows)
+            {
+                .fflush(fps); // before changing translation mode
+                fd = ._fileno(fps);
+                oldMode = ._setmode(fd, _O_BINARY);
+                version(DIGITAL_MARS_STDIO)
+                {
+                    import core.atomic;
+
+                    // @@@BUG@@@ 4243
+                    oldInfo = __fhnd_info[fd];
+                    atomicOp!"&="(__fhnd_info[fd], ~FHND_TEXT);
+                }
+            }
+        }
+
+    public:
+        ~this()
+        {
+            if(fps)
```

Comment:
Let me explain again my pet peeve about this "garden path" code structure (akin to https://en.wikipedia.org/wiki/Garden_path_sentence):

```d
void fun()
{
    if (condition)
    {
        several lines of statements
    }
   // no else, no code after if
}
```

With entering the `if`, the reader makes a mental note to remember that this code is conditional, then starts reading the code inside. At the end of the `if`, the reader figures that conditional code is done and is ready to either go through an `else` clause or some unconditional code that follows. But in fact there is no code following the `if` at all, which means the entire function's body _is_ the conditional. That's a much simpler structure and the reader must drop the mental note, now wasted effort, and reinterpret the function as a do-nothing unless the condition is met. The simpler and superior structure is:

```d
void fun()
{
    if (!condition)
        return;
    several lines of statements
}
```

Now the reader figures immediately that the function is a noop if the condition is met (most often a trivial case that deserves no attention). No more mental note, no more scoping, no more nonlinearity.